nemo/collections/asr/parts/nmse_clustering.py

```diff
@@ -0,0 +1,221 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This file is part of https://github.com/tango4j/Auto-Tuning-Spectral-Clustering.
+
+import numpy as np
+import scipy
+from scipy import sparse
+from sklearn.cluster import SpectralClustering as sklearn_SpectralClustering
+from sklearn.metrics.pairwise import cosine_similarity
+from sklearn.preprocessing import MinMaxScaler
+
+scaler = MinMaxScaler(feature_range=(0, 1))
+
+
+def get_kneighbors_conn(X_dist, p_neighbors):
+    X_dist_out = np.zeros_like(X_dist)
+    for i, line in enumerate(X_dist):
+        sorted_idx = np.argsort(line)
+        sorted_idx = sorted_idx[::-1]
+        indices = sorted_idx[:p_neighbors]
+        X_dist_out[indices, i] = 1
+    return X_dist_out
+
+
+def get_X_conn_from_dist(X_dist_raw, p_neighbors):
+    X_r = get_kneighbors_conn(X_dist_raw, p_neighbors)
+    X_conn_from_dist = 0.5 * (X_r + X_r.T)
+    return X_conn_from_dist
+
+
+def isFullyConnected(X_conn_from_dist):
+    gC = _graph_connected_component(X_conn_from_dist, 0).sum() == X_conn_from_dist.shape[0]
+    return gC
+
+
+def gc_thres_min_gc(mat, max_n, n_list):
+    p_neighbors, index = 1, 0
+    X_conn_from_dist = get_X_conn_from_dist(mat, p_neighbors)
+    fully_connected = isFullyConnected(X_conn_from_dist)
+    for i, p_neighbors in enumerate(n_list):
+        fully_connected = isFullyConnected(X_conn_from_dist)
+        X_conn_from_dist = get_X_conn_from_dist(mat, p_neighbors)
+        if fully_connected or p_neighbors > max_n:
+            break
+
+    return X_conn_from_dist, p_neighbors
+
+
+def _graph_connected_component(graph, node_id):
+    n_node = graph.shape[0]
```

Comment:
Add docstrings

Reply:
This and three more functions were taken from sklearn code. So left as is.

Reply:
Alright